src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectWorkspaceStateGenerator.cs

```diff
@@ -254,28 +185,95 @@ await _projectManager
         OnBackgroundWorkCompleted();
     }
 
-    private void OnStartingBackgroundWork()
+    /// <summary>
+    ///  Attempts to produce a <see cref="ProjectWorkspaceState"/> from the provide <see cref="Project"/> and <see cref="IProjectSnapshot"/>.
+    ///  Returns <see langword="null"/> if an error is encountered.
+    /// </summary>
+    private async Task<ProjectWorkspaceState?> GetProjectWorkspaceStateAsync(
+        Project? workspaceProject,
+        IProjectSnapshot projectSnapshot,
+        CancellationToken cancellationToken)
     {
-        if (BlockBackgroundWorkStart != null)
+        // This is the simplest case. If we don't have a project (likely because it is being removed),
+        // we just a default ProjectWorkspaceState.
+        if (workspaceProject is null)
+        {
+            return ProjectWorkspaceState.Default;
+        }
+
+        // Specifically not using BeginBlock because we want to capture cases where tag helper discovery never finishes.
+        var telemetryId = Guid.NewGuid();
+
+        _telemetryReporter.ReportEvent("taghelperresolve/begin", Severity.Normal,
+            new("id", telemetryId),
+            new("tagHelperCount", projectSnapshot.ProjectWorkspaceState.TagHelpers.Length));
+
+        try
+        {
+            var csharpLanguageVersion = workspaceProject.ParseOptions is CSharpParseOptions csharpParseOptions
+                ? csharpParseOptions.LanguageVersion
+                : LanguageVersion.Default;
+
+            using var _ = StopwatchPool.GetPooledObject(out var watch);
+
+            watch.Restart();
+            var tagHelpers = await _tagHelperResolver
+                .GetTagHelpersAsync(workspaceProject, projectSnapshot, cancellationToken)
+                .ConfigureAwait(false);
+            watch.Stop();
+
+            // don't report success if the work was cancelled
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
should this just return null instead of throwing?

Reply:
ah I see, you're catching below

Reply:
Yeah, that's what the code did before so I left it.